src/OVAL/oval_session.c

```diff
@@ -0,0 +1,657 @@
+	if (session->oval.variables) {
+		session->var_model = oval_variable_model_import_source(session->oval.variables);
+
+		if (session->var_model == NULL) {
+			oscap_seterr(OSCAP_EFAMILY_OVAL, "Failed to import the OVAL Variables "
+					"from '%s'.\n", session->oval.variables);
+			return 1;
+		}
+
+		if (oval_definition_model_bind_variable_model(session->def_model, session->var_model)) {
+			oscap_seterr(OSCAP_EFAMILY_OVAL, "Failed to bind Variables to Definitions\n");
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+static int oval_session_load_syschars(struct oval_session *session, xml_reporter fn)
+{
+	__attribute__nonnull__(session);
+
+	if (session->oval.syschars == NULL) {
+		oscap_seterr(OSCAP_EFAMILY_OVAL, "No OVAL System Characteristics set.\n");
+		return 1;
+	}
+	if (session->def_model == NULL) {
+		oscap_seterr(OSCAP_EFAMILY_OVAL, "No OVAL Definion loaded.\n");
+		return 1;
+	}
+	if (!oval_session_validate(session->oval.syschars, fn, OSCAP_DOCUMENT_OVAL_SYSCHAR)) {
+		return 1;
+	}
+
+	/* this will overwrite collect's sys model */
+	session->sys_model = oval_syschar_model_new(session->def_model);
+	if (oval_syschar_model_import_source(session->sys_model, session->oval.syschars) == -1) {
+		oscap_seterr(OSCAP_EFAMILY_OVAL, "Failed to import the System Characteristics from '%s'.\n",
+				oscap_source_readable_origin(session->oval.syschars));
+		oval_syschar_model_free(session->sys_model);
+	}
+
+	return 0;
+}
+
+int oval_session_load(struct oval_session *session, xml_reporter fn)
+{
+	__attribute__nonnull__(session);
+	__attribute__nonnull__(fn);
+
+	int ret = 0;
+
+	if ((ret = oval_session_load_definitions(session, fn)) != 0) {
+		return ret;
+	}
+	if (session->oval.variables && (ret = oval_session_load_variables(session, fn)) != 0) {
+		return ret;
+	}
+	if (session->oval.syschars && oval_session_load_syschars(session, fn) != 0) {
+		return ret;
+	}
+
+	return ret;
+}
+
+static int oval_session_setup_agent(struct oval_session *session)
+{
+	__attribute__nonnull__(session);
+
+	char *path_clone;
+
+	path_clone = oscap_strdup(oscap_source_readable_origin(session->oval.definitions));
+	if (path_clone == NULL) {
+		return 1;
+	}
+
+	/* free the previous Agent session if this function was already called */
+	if (session->sess) oval_agent_destroy_session(session->sess);
+
+	session->sess = oval_agent_new_session(session->def_model, basename(path_clone));
+	if (session->sess == NULL) {
+		oscap_seterr(OSCAP_EFAMILY_OVAL, "Failed to create a new agent session.");
+		free(path_clone);
+		return 1;
+	}
+	free(path_clone);
+
+	oval_agent_set_product_name(session->sess, (char *)oscap_productname);
+	return 0;
+}
+
+int oval_session_evaluate_id(struct oval_session *session, char *probe_root, const char *id, oval_result_t *result)
+{
+	__attribute__nonnull__(session);
+
+	if (probe_root) {
+		if (setenv("OSCAP_PROBE_ROOT", probe_root, 1) != 0) {
+			oscap_seterr(OSCAP_EFAMILY_OSCAP, "Failed to set the OSCAP_PROBE_ROOT environment variable.\n");
```

Comment:
Please remove trailing `\n`. oscap_seterr adds `[filename:linenum]\n` to the error message.

(multiple occurrences)